core/src/price_estimation/threaded_price_source.rs

```diff
@@ -1,90 +1,48 @@
+/// Implements a `PriceSource` that is always immediately ready by updating itself in a background
+/// task and using the most recent update.
 pub struct ThreadedPriceSource {
-    // Shared between this struct and the thread. The background thread writes
-    // the prices, the thread calling `get_prices` reads them.
+    // Shared between this struct and the task. The task writes the prices and `get_prices` reads
+    // them.
     price_map: Arc<Mutex<HashMap<TokenId, u128>>>,
-    // Allows the thread to notice when the owning struct is dropped.
-    // Mutex is not needed because we don't access the sender at all but we need
-    // this struct to be sync and this makes the compiler happy.
-    _channel_to_thread: Mutex<Sender<()>>,
-
-    // To make testing easier we use another channel to which the thread sends
-    // a message whenever it completes one update loop.
-    #[cfg(test)]
-    test_receiver: Mutex<Receiver<()>>,
 }
 
 impl ThreadedPriceSource {
     /// All token prices will be updated every `update_interval`. Prices for
     /// other tokens will not be returned in `get_prices`.
     ///
-    /// The join handle represents the background thread. It can be used to
-    /// verify that it exits when the created struct is dropped.
+    /// The join handle represents the task. It can be used to verify that it exits when the struct
+    /// is dropped.
     pub fn new<T: 'static + PriceSource + Send>(
         tokens: Vec<TokenId>,
         price_source: T,
         update_interval: Duration,
     ) -> (Self, JoinHandle<()>) {
         let price_map = Arc::new(Mutex::new(HashMap::new()));
-        let (sender, receiver) = mpsc::channel::<()>();
-
-        #[cfg(test)]
-        let (test_sender, test_receiver) = mpsc::channel::<()>();
-
-        let join_handle = thread::spawn({
-            let price_map = price_map.clone();
-
-            // vk: Clippy notes that the following loop and match can be written
-            // as a `while let` loop but I find it clearer to make all cases
-            // explicit.
-            #[allow(clippy::while_let_loop)]
-            move || loop {
-                match receiver.recv_timeout(update_interval) {
-                    Ok(()) | Err(RecvTimeoutError::Timeout) => {
-                        // Make sure we don't hold the mutex while this blocking call happens.
-                        match price_source.get_prices(&tokens).wait() {
-                            Ok(prices) => price_map.lock().wait().extend(prices),
-                            Err(err) => log::warn!("price_source::get_prices failed: {}", err),
-                        }
+        let join_handle = task::spawn({
+            let price_map = Arc::downgrade(&price_map);
+            async move {
+                while let Some(price_map) = price_map.upgrade() {
+                    match price_source.get_prices(&tokens).wait() {
+                        Ok(prices) => price_map.lock().wait().extend(prices),
```

Comment:
Does MutexGuard automatically convert to a Promis or what 🧙is happening here? Can this also be an await?